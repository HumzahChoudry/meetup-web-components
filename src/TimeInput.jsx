import React from 'react';
import cx from 'classnames';

/**
* @module TimeInput
*/
class TimeInput extends React.Component {

	constructor(props) {
		super(props);
		this.state = {
			value: this.props.value || ''
		};
		this.onChange = this.onChange.bind(this);
	}

	onChange(e) {
		this.setState({ value: e.target.value });
		this.props.onChangeCallback && this.props.onChangeCallback(e.target.value);
	}

	render() {
		const {
			onChangeCallback,	// eslint-disable-line no-unused-vars
			id,
			label,
			name,
			className,
			error,
			required,
			value,		// eslint-disable-line no-unused-vars
			...other
		} = this.props;

		const classNames = cx(
			'input--time',
			className
		);

		const labelClassNames = cx({ required });
		return (
			<span>
				{ label && <label htmlFor={id} className={labelClassNames}>{label}</label> }
				<input
					id={id}
					type='time'
					name={name}
					value={this.state.value}
					className={classNames}
					onChange={this.onChange}
					required={required}
					ref={ input => this.inputEl = input }
					{...other}
				/>
				{ error && <p className='text--error'>{error}</p> }
			</span>
		);

	}
}

TimeInput.propTypes = {
	onChangeCallback: React.PropTypes.func,
	name: React.PropTypes.string.isRequired,
	error: React.PropTypes.string,
	label: React.PropTypes.oneOfType([
		React.PropTypes.string,
		React.PropTypes.element
	]),
	required: React.PropTypes.bool
};

export default TimeInput;

