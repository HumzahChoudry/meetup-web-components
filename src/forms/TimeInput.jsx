import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

/**
* @module TimeInput
*/
class TimeInput extends React.Component {

	constructor(props) {
		super(props);

		this.onChange = this.onChange.bind(this);
	}

	/**
	* @function onChange
	* @param e Event Object
	* @description called when the input changes, in turn calls the onChange
	* 	handler prop, if there is one (eg supplied by redux-form) and an onChangeCallback
	*	which may be provided by a parent component such as DateTimePicker
	*/
	onChange(e) {
		this.props.onChange && this.props.onChange(e.target.value); // redux-form provides an onChange prop

		// onChangeCallback currently provided by a parent component eg DateTimePicker
		this.props.onChangeCallback && this.props.onChangeCallback(e.target.value);
	}

	render() {
		const {
			id,
			label,
			name,
			className,
			required,
			value,
			error,
			onChange,			// eslint-disable-line no-unused-vars
			onChangeCallback, 	// eslint-disable-line no-unused-vars
			...other
		} = this.props;

		const classNames = cx(
			'input--time',
			className
		);

		const labelClassNames = cx(
			'label--field',
			{ required }
		);

		const errorId = `${id}-error`;

		if (error) {
			other['aria-invalid'] = true;
			other['aria-describedby'] = errorId;
		}

		return (
			<span>
				{ label && <label htmlFor={id} className={labelClassNames}>{label}</label> }
				<input
					id={id}
					type='time'
					name={name}
					value={value}
					className={classNames}
					required={required}
					onChange={this.onChange}
					ref={ input => this.inputEl = input }
					{...other}
				/>
				{ error && <p id={errorId} className='text--error'>{error}</p> }
			</span>
		);

	}
}

TimeInput.propTypes = {
	name: PropTypes.string.isRequired,
	error: PropTypes.string,
	label: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element
	]),
	required: PropTypes.bool,
	onChange: PropTypes.func,			// redux-form provides an onChange prop
	onChangeCallback: PropTypes.func, 	// currently provided by a parent component eg DateTimePicker
};

export default TimeInput;

