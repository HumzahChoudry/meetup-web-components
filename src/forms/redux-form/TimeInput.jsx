import React from 'react';
import TimeInput from '../TimeInput';

/**
 * This component wraps the standard TimeInput for use with redux-form.
 * It deconstructs props that redux-form sets and re-sets them on TimeInput.
 * @param {object} props React component props
 * @return {React.Component} TimeInput
 */
const ReduxFormTimeInput = props => {
	const { input, meta, ...other } = props;
	console.log('HELLO HELLO HELLO');
	return <TimeInput {...input} error={meta.touched && meta.error} {...other} />;
};

ReduxFormTimeInput.displayName = 'ReduxFormTimeInput';

export default ReduxFormTimeInput;
