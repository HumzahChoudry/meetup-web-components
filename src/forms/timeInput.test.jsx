
import React from 'react';
import { mount } from 'enzyme';
import TimeInput, {
	HOURS_INPUT_CLASS,
	MINUTES_INPUT_CLASS,
	MERIDIAN_INPUT_CLASS,
	getTimeParts
} from './TimeInput';


describe('TimeInput', function() {

	let component,
		component12Hr,
		hoursInputEl,
		minutesInputEl,
		inputEl,
		onChangeSpy,
		onNumberChangeSpy,
		onNumberMouseupSpy,
		onMeridianChangeSpy,
		onTimeInputChangeSpy;

	const onChangePropMock = jest.fn(),
		newTime = '23:00',
		newHour = '23',
		hourValue = '22',
		minuteValue = '12',
		overMax = '666',
		underMax = '-666';

	const props = {
		name: 'time',
		value: `${hourValue}:${minuteValue}`,
		onChange: onChangePropMock,
		required: true
	};

	describe('TimeInput, basic functionality', () => {

		describe('getTimeParts', () => {
			it('returns an object of hours, min, meridian from a time string', () => {
				const result = getTimeParts('13:00');
				const expected = {
					hours: '13',
					minutes: '00',
					meridian: 'PM'
				};
				expect(result).toEqual(expected);
			});
		});

		describe('parseValueForState', () => {
			it('takes a value string for 24Hr time and returns an object with hours, min, meridian, value', () => {
				const value = '13:00',
				is24Hr = true;
				const result = TimeInput.prototype.parseValueIntoState(value, is24Hr);
				const expected = {
					hours: '13',
					minutes: '00',
					meridian: false,
					value
				};
				expect(result).toEqual(expected);
			});
			it('takes a value string, is24hr as false, and returns an object with hours, min, meridian, value', () => {
				const value = '13:15',
					is24Hr = false;
				const result = TimeInput.prototype.parseValueIntoState(value, is24Hr);
				const expected = {
					hours: '01',
					minutes: '15',
					meridian: 'PM',
					value
				};
				expect(result).toEqual(expected);
			});
		});

		describe('constrainValue', () => {
			it('returns the same value if no min and max', () => {
				const value = '10';
				const result = TimeInput.prototype.constrainValue(undefined, undefined, value);
				expect(result).toEqual(parseInt(value));
			});
			it('returns the same value if within min and max', () => {
				const value = '10';
				const result = TimeInput.prototype.constrainValue(2, 22, value);
				expect(result).toEqual(parseInt(value));
			});
			it('returns the constrained value if min', () => {
				const value = '10';
				const min = 20;
				const result = TimeInput.prototype.constrainValue(min, undefined, value);
				expect(result).toEqual(min);
			});
			it('returns the constrained value if max', () => {
				const value = '10';
				const max = 5;
				const result = TimeInput.prototype.constrainValue(undefined, max, value);
				expect(result).toEqual(max);
			});
		});
	});

	describe('TimeInput, with input[time] support', () => {
		beforeEach(() => {
			onChangeSpy = jest.spyOn(TimeInput.prototype, 'onChange');
			onTimeInputChangeSpy = jest.spyOn(TimeInput.prototype, 'onTimeInputChange');
			component = mount(<TimeInput {...props} />);
			inputEl = component.find('input');
		});

		afterEach(() => {
			component = null;
			inputEl = null;
			onChangeSpy.mockClear();
			onTimeInputChangeSpy.mockClear();
		});

		it('renders a time html input with expected props', function() {
			expect(component).toMatchSnapshot();
		});

		it('calls onTimeInputChange when value is changed', function() {
			inputEl.simulate('change', { target: { value: newTime } });
			expect(onTimeInputChangeSpy).toHaveBeenCalled();
		});

		it('calls onChange prop when value is changed', function() {
			component.instance().onChange(newTime);
			expect(onChangePropMock).toHaveBeenCalledWith(newTime);
		});

	});

	describe('TimeInput, without input[time] support', () => {
		beforeEach(() => {
			onChangeSpy = jest.spyOn(TimeInput.prototype, 'onChange');
			onNumberChangeSpy = jest.spyOn(TimeInput.prototype, 'onNumberChange');
			onNumberMouseupSpy = jest.spyOn(TimeInput.prototype, 'onNumberMouseup');
			onMeridianChangeSpy = jest.spyOn(TimeInput.prototype, 'onMeridianChange');

			component = mount(<TimeInput {...props} />);
			component.setState({supportsTime: false});

			component12Hr = mount(<TimeInput is24Hr={false} {...props} />);
			component12Hr.setState({supportsTime: false});

			hoursInputEl = component.find(`.${HOURS_INPUT_CLASS}`);
			minutesInputEl = component.find(`.${MINUTES_INPUT_CLASS}`);
		});

		afterEach(() => {
			component = null;
			hoursInputEl = null;
			minutesInputEl = null;
			onChangeSpy.mockClear();
			onNumberChangeSpy.mockClear();
			onNumberMouseupSpy.mockClear();
		});

		it('calls internal onChange when value is changed', () => {
			hoursInputEl.simulate('change', { target: { value: newHour } });
			hoursInputEl.simulate('blur');

			expect(onChangeSpy).toHaveBeenCalled();
		});

		it('calls onChange prop when value is changed', () => {
			component.instance().onChange(newHour);
			hoursInputEl.simulate('blur');

			expect(onChangePropMock).toHaveBeenCalledWith(newTime);
		});

		it('calls onChange prop with appropriately formatted time', () => {
			component.setState(component.instance().parseValueIntoState('15:15', true));
			onChangePropMock.mockClear();

			hoursInputEl.instance().value = '1';
			hoursInputEl.simulate('change');
			hoursInputEl.simulate('blur');

			expect(onChangePropMock).toHaveBeenCalledWith('01:15');
		});

		it('calls onChange prop with appropriately formatted time, hours from 12 into 24Hr', () => {
			const hoursInputEl_12Hr = component12Hr.find(`.${HOURS_INPUT_CLASS}`);
			const meridianInputEl = component12Hr.find(`select.${MERIDIAN_INPUT_CLASS}`);

			component12Hr.setState(component.instance().parseValueIntoState('15:15', false));
			onChangePropMock.mockClear();

			meridianInputEl.instance().value = 'PM';
			meridianInputEl.simulate('change');

			hoursInputEl_12Hr.instance().value = '1';
			hoursInputEl_12Hr.simulate('change');
			hoursInputEl_12Hr.simulate('blur');

			expect(onChangePropMock).toHaveBeenCalledWith('13:15');
		});

		it('renders the meridian input for 12hr time', function() {
			const meridianInputEl = component12Hr.find(`select.${MERIDIAN_INPUT_CLASS}`);
			meridianInputEl.simulate('change', { target: { value: 'PM' } });

			expect(onMeridianChangeSpy).toHaveBeenCalled();
		});

		it('does not render meridian input for 24hr time', function() {
			expect(component.find(`select.${MERIDIAN_INPUT_CLASS}`).length).toBe(0);
		});

		it('does not accept hour values outside the minimum or maximum', function() {
			hoursInputEl.instance().value = overMax;
			hoursInputEl.simulate('change');
			hoursInputEl.simulate('blur');

			expect(parseInt(hoursInputEl.instance().value, 10)).toEqual(hoursInputEl.prop('max'));

			hoursInputEl.instance().value = underMax;
			hoursInputEl.simulate('change');
			hoursInputEl.simulate('blur');

			expect(parseInt(hoursInputEl.instance().value, 10)).toEqual(hoursInputEl.prop('min'));
		});

		it('does not accept minute values outside the minimum or maximum', function() {
			minutesInputEl.instance().value = overMax;
			minutesInputEl.simulate('change');
			minutesInputEl.simulate('blur');

			expect(parseInt(minutesInputEl.instance().value, 10)).toEqual(minutesInputEl.prop('max'));

			minutesInputEl.instance().value = underMax;
			minutesInputEl.simulate('change');
			minutesInputEl.simulate('blur');

			expect(parseInt(minutesInputEl.instance().value, 10)).toEqual(minutesInputEl.prop('min'));
		});

		it('should update time value when receiving hours or minutes prop', () => {
			const instance = component.instance();
			const newProps = { value: '23:00', is24Hr: true };
			const expected = TimeInput.prototype.parseValueIntoState(newProps.value, newProps.is24Hr);
			jest.spyOn(instance, 'setState');
			instance.componentWillReceiveProps(newProps);
			expect(instance.setState).toHaveBeenCalledWith(expected);
		});

		it('should increase hours or minutes value when pressing up key', () => {
			const newHourValue = (parseInt(hourValue, 10) + 1);
			const newMinuteValue = (parseInt(minuteValue, 10) + 1);

			expect(parseInt(hoursInputEl.instance().value, 10)).toEqual(parseInt(hourValue));
			hoursInputEl.simulate('keyDown', {keyCode: 38});
			expect(parseInt(hoursInputEl.instance().value, 10)).toEqual(parseInt(newHourValue));

			expect(parseInt(minutesInputEl.instance().value, 10)).toEqual(parseInt(minuteValue));
			minutesInputEl.simulate('keyDown', {keyCode: 38});
			expect(parseInt(minutesInputEl.instance().value, 10)).toEqual(parseInt(newMinuteValue));
		});

		it('should select hours or minutes input text when clicking', () => {
			hoursInputEl.simulate('mouseUp');
			expect(onNumberMouseupSpy).toHaveBeenCalled();
		});

	});

});

