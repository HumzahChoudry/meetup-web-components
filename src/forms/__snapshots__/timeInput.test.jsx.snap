// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeInput TimeInput, with input[time] support renders a time html input with expected props 1`] = `
<InputTime
  name="time"
  onChange={[Function]}
  required="*"
  value="11:15"
/>
`;
